File: models/FuelConsumption.js
// models/FuelConsumption.js
const { Schema, model, Types } = require('mongoose');

const fuelConsumptionSchema = new Schema({
    organizationId: { type: Types.ObjectId, ref: 'Organization', required: true },
    period: { year: { type: Number, required: true }, month: { type: Number, required: true, min: 1, max: 12 } },
    fuelType: { type: String, enum: ['diesel', 'gasolina', 'glp', 'gln', 'queroseno', 'otro'], required: true },
    quantity: { type: Number, required: true, min: 0 },
    unit: { type: String, enum: ['L', 'gal', 'kg', 'm3'], required: true },
    emissionFactor: {
        value: { type: Number, required: true, min: 0 }, // kgCO2e por unidad
        unit: { type: String, default: 'kgCO2e/unit' },
        source: { type: String, default: 'Oficial/IPCC' },
        year: { type: Number }
    },
    co2e: { type: Number, required: true, min: 0 }, // quantity * factor
    scope: { type: String, default: 'scope1' },
    source: { type: String, enum: ['manual', 'receipt', 'telematics'], default: 'manual' },
    receiptId: { type: Types.ObjectId, ref: 'Receipt' }
}, { timestamps: true });

fuelConsumptionSchema.index({ organizationId: 1, 'period.year': 1, 'period.month': 1 });
fuelConsumptionSchema.index({ fuelType: 1 });

module.exports = model('FuelConsumption', fuelConsumptionSchema);

File: Domain/FuelConsumption.cs
using Ecotrack_Api.Domain.Common;
using MongoDB.Bson;
using MongoDB.Bson.Serialization.Attributes;


namespace Ecotrack_Api.Domain;


public class FuelConsumption : BaseEntity
{
    [BsonRepresentation(BsonType.ObjectId)]
    public string OrganizationId { get; set; } = string.Empty;
    public Period Period { get; set; } = new();


    public string FuelType { get; set; } = string.Empty; // gasolina, diésel, etc.
    public decimal Quantity { get; set; }
    public string Unit { get; set; } = "L"; // L | gal | kg | m3


    public EmissionFactor EmissionFactor { get; set; } = new();
    public decimal Co2e { get; set; }
    public string Scope { get; set; } = "scope1";
    public string Source { get; set; } = "manual"; // manual | receipt | telematics


    [BsonRepresentation(BsonType.ObjectId)]
    public string? ReceiptId { get; set; }
}